File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py

from heapq import *

def solution(book_time):
    book_time.sort()
    heap = []

    print(book_time)
    last = -1
    cnt= 0
    for time in book_time:
        start_hour = time[0].split(':')[0]
        start_min = time[0].split(':')[1]
        end_hour = time[1].split(':')[0]
        end_min = time[1].split(':')[1]


        start = 60*int(start_hour) + int(start_min)
        end = 60 * int(end_hour) + int(end_min)

        if heap:
            last = heappop(heap)
            if last+10<=start:
                heappush(heap,end )
            else:
                heappush(heap, end )
                cnt+=1
                heappush(heap,last)
        else:
            heappush(heap, end)
            cnt += 1

    return cnt

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.Arrays;
import java.util.PriorityQueue;
class Solution {
    public int solution(String[][] book_time) {
        int cnt = 0;
        
        Arrays.sort(book_time, (o1,o2)-> o1[0].compareTo(o2[0]));
        
        
        System.out.println(Arrays.deepToString(book_time));

        
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        int last = -1;
        
        for(String[] time : book_time) {
        	String start_hour = time[0].split(":")[0];
        	String start_min = time[0].split(":")[1];
        	String end_hour = time[1].split(":")[0];
        	String end_min = time[1].split(":")[1];
        	
        	int start = 60 * Integer.parseInt(start_hour) + Integer.parseInt(start_min);
        	int end = 60 * Integer.parseInt(end_hour) + Integer.parseInt(end_min);
        	
        	if(!heap.isEmpty()) {
        		last = heap.poll();
        		if(last+10<=start) {
        			heap.add(end);
        		}else {
        			heap.add(end);
        			cnt++;
        			heap.add(last);
        		}
        		
        	}else {
        		heap.add(end);
        		cnt++;
        	}
        	
        	
        	
        }
        		
        
        
        
        
        
        
		
        return cnt;
    }
}